### ISSUE TITLE
Error: "Nesting level too deep - recursive dependency?"

### ISSUE THREAD

**Issue by haralake** [OFFLINE-GmbH/oc-mall-plugin#506]
I created 4 level categories product that belongs on 2 of those subcategories. When i try to open product i get error:
`Nesting level too deep - recursive dependency?`

/../../../../public/vendor/laravel/framework/src/Illuminate/Support/Collection.php line 1515

```
     * @param  bool  $strict
     * @return static
     */
    public function unique($key = null, $strict = false)
    {
        if (is_null($key)) {
1515->            return new static(array_unique($this->items, SORT_REGULAR));
        }
  
        $callback = $this->valueRetriever($key);
  
        $exists = [];
```

**Comment by tobias-kuendig** [comment]
Do you have a full stack trace?

**Comment by haralake** [comment]
Symfony\Component\Debug\Exception\FatalErrorException: Nesting level too deep - recursive dependency? in /../../../../public/vendor/laravel/framework/src/Illuminate/Support/Collection.php:1515

Stack trace:
#0 {main}

**Comment by tobias-kuendig** [comment]
Is this all there is? Are there no more lines below this? Also check your system.log for more details. I need to know where the error originates from, otherwise it's hard to debug this problem.

What action do you make before this error appears, what page do you visit?

**Comment by haralake** [comment]
Error log on Octobercms gives only this. I also checked server error log and only this exact error appears.
On the backend i open a product that belongs on a category inside one category that is inside one other category (3 level deep). This product has checked only the last category that belongs (on categories field). So i also checked the category that category belongs, and save changes. Then when i open product (using product page) on frontend i get this error.

**Comment by haralake** [comment]
The weird thing is that if i check the first and the last category is working without error. But if i check the middle one i get the error. I really don't understand..

**Comment by tobias-kuendig** [comment]
Alright, this definitely sounds like an infinite loop. I'll see if I can reproduce it on my end.

**Comment by tobias-kuendig** [comment]
Are you able to reproduce this issue on our demo website? I could not reproduce it so far.

It looks like the problem is a circular dependency, that causes PHP to throw the error when using `array_` functions. It could also depend on a PHP setting, when this error is thrown (how many recursions are allowed).

**Comment by haralake** [comment]
Yes i just reproduce it on your demo site and it seems that this error is connected with product variants. Steps to reproduce it:
1. Create a category "test" inside category "Bikes", create category "test2" inside category "test", create category "test3" inside category "test2"
2. Check "Inherit properties of parent category" on categories you created
3. Create a product variant using "Attribute for variant grouping" for example.
4. Check categories "test3" and "test" for this product.
5. Save it
6. Now if you open product variant on frontend you get error "Nesting level too deep - recursive dependency?"

**Comment by tobias-kuendig** [comment]
Thank you very much for your effort here! I'll take a look at this.

**Comment by tobias-kuendig** [comment]
Thanks again @haralake, this is fixed in 0c6408f!
